File: integration-test/src/test/java/com/olx/ps/vault/LogicalExtensionsTest.java
package com.olx.ps.vault;

import com.bettercloud.vault.Vault;
import com.bettercloud.vault.VaultException;
import com.olx.ps.testContainer.VaultContainer;
import org.junit.BeforeClass;
import org.junit.ClassRule;
import org.junit.Test;

import java.io.IOException;
import java.time.Duration;

import static io.kotlintest.matchers.MatchersKt.shouldEqual;
import static java.util.Collections.singletonMap;

public class LogicalExtensionsTest {

    @ClassRule
    public static VaultContainer container = new VaultContainer();

    @BeforeClass
    public static void beforeClass()
            throws IOException, InterruptedException, VaultException {
        container.initAndUnsealVault();
        container.setupBackendUserPass();
        container.getVault()
                .auth()
                .loginByUserPass(VaultContainer.USER_ID, VaultContainer.PASSWORD);
    }

    @Test
    public void construct_data_class_from_response() throws VaultException {
        Vault vault = container.getRootVault();
        vault.logical().write("secret/hello", singletonMap("secret", "hello"));
        GenericCredentials expected = new GenericCredentials("hello", Duration.ofHours(24));

        GenericCredentials actual = LogicalExtensions.read(
                vault.logical(),
                "secret/hello",
                GenericCredentials::new
        );
        shouldEqual(actual, expected);
    }
}

File: vault-kotlin/src/test/kotlin/com/olx/ps/vault/LogicalExtensionsTest.kt
package com.olx.ps.vault

import com.olx.ps.vault.utils.GenericCredentials
import com.olx.ps.vault.utils.MockVault
import com.olx.ps.vault.utils.secretResponse
import com.olx.ps.vault.utils.server
import io.kotlintest.matchers.shouldEqual
import org.junit.Test
import java.time.Duration

class LogicalExtensionsTest {

    @Test
    fun `construct data class from LogicalResponse`() {
        val handler = MockVault(mutableListOf(secretResponse to 200))
        server(8200, handler) {
            val vault = vault {
                config {
                    address("http://localhost:8200")
                    token("E592D3F7-640C-4CF7-AA4F-40547E97A3CD")
                    openTimeout(5)
                    readTimeout(10)
                }
            }
            val expected = GenericCredentials("secret1", Duration.ofHours(768))
            val actual = vault.logical().read("secret/hello", ::GenericCredentials)
            expected shouldEqual actual
        }
    }
}
